core/src/ecmult.rs

```diff
@@ -9,18 +13,149 @@ pub const ECMULT_TABLE_SIZE_A: usize = 1 << (WINDOW_A - 2);
 
+fn odd_multiples_table_storage_var(
+    pre: &mut [AffineStorage],
+    a: &Jacobian
+) {
+    let mut prej: Vec<Jacobian> = Vec::with_capacity(pre.len());
+    for _ in 0..pre.len() {
+        prej.push(Jacobian::default());
+    }
+    let mut prea: Vec<Affine> = Vec::with_capacity(pre.len());
+    for _ in 0..pre.len() {
+        prea.push(Affine::default());
+    }
+    let mut zr: Vec<Field> = Vec::with_capacity(pre.len());
+    for _ in 0..pre.len() {
+        zr.push(Field::default());
+    }
+
+    odd_multiples_table(&mut prej, &mut zr, a);
+    set_table_gej_var(&mut prea, &prej, &zr);
+
+    for i in 0..pre.len() {
+        pre[i] = prea[i].clone().into();
+    }
+}
+
 /// Context for accelerating the computation of a*P + b*G.
 pub struct ECMultContext {
     pre_g: [AffineStorage; ECMULT_TABLE_SIZE_G],
 }
 
 impl ECMultContext {
-    /// Create a new `ECMultContext`.
-    pub const fn new(pre_g: [AffineStorage; ECMULT_TABLE_SIZE_G]) -> Self {
+    /// Create a new `ECMultContext` from raw values.
+    pub const unsafe fn new_raw(pre_g: [AffineStorage; ECMULT_TABLE_SIZE_G]) -> Self {
         Self { pre_g }
     }
+
+    /// Inspect raw values of `ECMultContext`.
+    pub fn inspect_raw(&self) -> &[AffineStorage; ECMULT_TABLE_SIZE_G] {
+        &self.pre_g
+    }
+
+    /// Deconstruct the `ECMultContext` into raw `AffineStorage`.
+    pub fn deconstruct_raw(self) -> [AffineStorage; ECMULT_TABLE_SIZE_G] {
+        self.pre_g
+    }
+
+    /// Generate a new `ECMultContext` in the heap. Note that this function is expensive.
+    pub fn new_boxed() -> Box<Self> {
+        let mut this = unsafe {
+            let ptr = alloc(Layout::new::<ECMultContext>()) as *mut ECMultContext;
+            let mut this = Box::from_raw(ptr);
+
+            for i in 0..ECMULT_TABLE_SIZE_G {
+                this.pre_g[i] = AffineStorage::default();
+            }
+
+            this
+        };
+
+        let mut gj = Jacobian::default();
+        gj.set_ge(&AFFINE_G);
+        odd_multiples_table_storage_var(&mut this.pre_g, &gj);
+
+        this
+    }
 }
 
+pub fn set_all_gej_var(a: &[Jacobian]) -> Vec<Affine> {
+    let mut az: Vec<Field> = Vec::with_capacity(a.len());
+    for i in 0..a.len() {
+        if !a[i].is_infinity() {
+            az.push(a[i].z.clone());
+        }
+    }
+    let azi: Vec<Field> = inv_all_var(&az);
+
+    let mut ret = Vec::with_capacity(a.len());
+    for _ in 0..a.len() {
+        ret.push(Affine::default());
+    }
+
+    let mut count = 0;
+    for i in 0..a.len() {
+        ret[i].infinity = a[i].infinity;
+        if !a[i].is_infinity() {
+            ret[i].set_gej_zinv(&a[i], &azi[count]);
+            count += 1;
+        }
+    }
+    ret
+}
+
+/// Calculate the (modular) inverses of a batch of field
+/// elements. Requires the inputs' magnitudes to be at most 8. The
+/// output magnitudes are 1 (but not guaranteed to be
+/// normalized). The inputs and outputs must not overlap in
+/// memory.
```

Comment:
what do you mean by "The inputs and outputs must not overlap in memory"? outputs are not supplied by a user here?

Reply:
I think that was from C library and does not really apply here. Let me remove it.